### ISSUE TITLE
[gcov] Reduced coverage data on first run

### ISSUE THREAD

**Issue by IvanRibakov** [ThrowTheSwitch/Ceedling#442]
After updating from c5e2c90cd94 to recent version of Ceedling (v0.29.1, 91b4ed2) our CI job that checks code coverage started reporting significantly smaller overall coverage (line coverage went down from 70% to 30%).

While debugging the issue locally I noticed that smaller coverage is only reported on a first run (no previous test build data available). Subsequent runs however are still reporting correct number. Deleting test build folder (or running `ceedling clobber`) and then rerunning tests with coverage reliably reproduces incorrect coverage data.

Unfortunately, I only observe this behaviour when running entire test suite (`gcov:all`). Selecting one of the files that appears uncovered on the first run and running a dedicated test suite seems to produce correct coverage for that specific source file.

Below are log lines for one select source file when running `ceedling gcov:all` on the first and second attempts:

- First attempt:

```
$ ceedling gcov:all

...
Test 'test_event_controller.c'
------------------------------
Generating runner for test_event_controller.c...
Compiling test_event_controller_runner.c...
Compiling test_event_controller.c...
Linking test_event_controller.out...
/usr/bin/ld: warning: libcrypto.so.1.0.0, needed by reqs/lib/libssl.so, may conflict with libcrypto.so.10
/usr/bin/ld: warning: libssl.so.10, needed by reqs/lib/libcurl.so, may conflict with libssl.so.1.0.0
Running test_event_controller.out...

...

--------------------------
GCOV: OVERALL TEST SUMMARY
--------------------------
TESTED:  434
PASSED:  434
FAILED:    0
IGNORED:   0


---------------------------
GCOV: CODE COVERAGE SUMMARY
---------------------------

...

event_controller.c Lines executed:0.00% of 388
event_controller.c Branches executed:0.00% of 152
event_controller.c Taken at least once:0.00% of 152
event_controller.c Calls executed:0.00% of 205

...
```

- Second attempt

```
$ ceedling gcov:all

...
Test 'test_event_controller.c'
------------------------------
Linking test_event_controller.out...
/usr/bin/ld: warning: libcrypto.so.1.0.0, needed by reqs/lib/libssl.so, may conflict with libcrypto.so.10
/usr/bin/ld: warning: libssl.so.10, needed by reqs/lib/libcurl.so, may conflict with libssl.so.1.0.0
Running test_event_controller.out...

...

--------------------------
GCOV: OVERALL TEST SUMMARY
--------------------------
TESTED:  434
PASSED:  434
FAILED:    0
IGNORED:   0


---------------------------
GCOV: CODE COVERAGE SUMMARY
---------------------------

...

event_controller.c Lines executed:44.07% of 388
event_controller.c Branches executed:55.26% of 152
event_controller.c Taken at least once:31.58% of 152
event_controller.c Calls executed:45.37% of 205

...
```

Any ideas what might be causing such behaviour or how could I proceed with debugging the root cause?

**Comment by CezaryGapinski** [comment]
Based on your log from the first attempt:
```
Test 'test_event_controller.c'
------------------------------
Generating runner for test_event_controller.c...
Compiling test_event_controller_runner.c...
Compiling test_event_controller.c...
Linking test_event_controller.out...
/usr/bin/ld: warning: libcrypto.so.1.0.0, needed by reqs/lib/libssl.so, may conflict with libcrypto.so.10
/usr/bin/ld: warning: libssl.so.10, needed by reqs/lib/libcurl.so, may conflict with libssl.so.1.0.0
Running test_event_controller.out...
```
`event_controller.c` file is not on the list. It means this file wasn't compiled and only linked to `test_event_controller.out`. I suppose the `event_controller.c` was compiled for the some other test and you also should check issue for this test where compilation of this source file occurs.

> After updating from c5e2c90 to recent version of Ceedling (v0.29.1, 91b4ed2) our CI job that checks code coverage started reporting significantly smaller overall coverage (line coverage went down from 70% to 30%).

The log list looks the same for the previous Ceedling version?
Do you use some specific Ceedling features like `auto_link_deep_dependencies` or your `event_controller.c` source file is surrounded by preprocessor ifdefs?

**Comment by tobyjwebb** [comment]
Hi @CezaryGapinski 

> The log list looks the same for the previous Ceedling version?

Currently, this would be rather difficult for us to check, but I'll look into it.

> Do you use some specific Ceedling features like auto_link_deep_dependencies or your event_controller.c source file is surrounded by preprocessor ifdefs?

Yes, we do use ``auto_link_deep_dependencies``. 
No, can't see any ifdefs in event_controller.c

Thanks for your feedback :)

**Comment by ojohnny** [comment]
I believe I have the same issue, and here is what I think happens for me:

1. The files which have lower coverage (or even 0%) are always files which are included in more than one test.
2. On the first run, the files are "compiled with coverage" multiple times, once for each test it is included from.  I guess whenever you recompile the object file, the coverage results resets?
3. On the second run, the file is compiled only once, and now the coverage is accumulated from all the tests where the file is included from.  Why it is even recompiled in this case is beyond me, I would have assumed it could reuse the compiled file from the first run.

So in my case it seems like this is a symptom of a bug "incremental build is not working on first run"?

I'm using Ceedling 0.29.1.

**Comment by IvanRibakov** [comment]
@CezaryGapinski Sorry for abandoning this ticket for quite some time. I finally had time to take another look at it. You asked some good questions [earlier](https://github.com/ThrowTheSwitch/Ceedling/issues/442#issuecomment-568729662) so I decided to try to reproduce the problem on a simpler setup. Here's what I got:

- Simplified source code to reproduce the issue: https://github.com/IvanRibakov/ceedling_bug_demo 

    - 3 source files: `a`, `b` and `c`.
        - Dependencies: `b <= a`, `c <= [a, b]`
    - 3 tests, one for each of the source files

       - `test_a <= a`
       - `test_b <= [a, b]`
       - `test_c <= [a, b, c]`

    - **Not using** "auto-link deep dependencies" feature - all includes are listed in the source + test files explicitly
    - Tried both with and without `use_test_preprocessor` - was able to reproduce odd behaviour **in both cases**

- Running `ceedling clobber test:all` multiple times behaves as expected - on each run each source file is compiled only once and each test is reusing previously compiled objects if available

- Running `ceedling clobber gcov:all utils:gcov` multiple times _sometimes_ results in incorrect coverage data because some source files (in case of this example it's `a.c`) are re-compiled. I say _sometimes_ because I'm unable to figure out the pattern that is causing recompilation - sometimes it happens on the first run and the subsequent runs are ok. Sometimes it would first produce good results and the suddenly recompile already compiled object.

To illustrate what I'm talking about, here's console output of 2 consecutive Ceedling runs. FYI - no source code was changed between consecutive runs in below sample output, although sometimes it helps to trigger the undesired behaviour by modifying include statement in `test_c.c`(`#include "a.h"` => `#include "basic/a.h"`)

<details>
  <summary>Run 1: Full console output</summary>

Notice how `Compiling a.c with coverage` appears only once in the logs and reported coverage for `a.c` is 100% at the end.

```
$ ceedling clobber gcov:all utils:gcov

Clobbering all generated files...
(For large projects, this task may take a long time to complete)



Test 'test_a.c'
---------------
Generating runner for test_a.c...
Compiling test_a_runner.c...
Compiling test_a.c...
Compiling unity.c...
Compiling a.c with coverage...
Compiling cmock.c...
Linking test_a.out...
Running test_a.out...


Test 'test_b.c'
---------------
Generating runner for test_b.c...
Compiling test_b_runner.c...
Compiling test_b.c...
Compiling b.c with coverage...
Linking test_b.out...
Running test_b.out...


Test 'test_c.c'
---------------
Generating runner for test_c.c...
Compiling test_c_runner.c...
Compiling test_c.c...
Compiling c.c with coverage...
Linking test_c.out...
Running test_c.out...
In your project.yml, define:

:gcov:
  :xml_report:

 to true or false to refine this feature.
For now, assumimg you do not want an xml report generated.
Creating a detailed html report of gcov results in build/artifacts/gcov/GcovCoverageResults.html...
Done.

-----------------
GCOV: TEST OUTPUT
-----------------
[test_a.c]
  - "a1() = 10"
  - "a2() = 10"
  - "a3() = 10"

[test_b.c]
  - "a1() = 10"
  - "b1() = 10"
  - "b2() = 10"
  - "b3() = 10"

[test_c.c]
  - "a1() = 10"
  - "b1() = 10"
  - "c1() = 10"
  - "c2() = 10"
  - "c3() = 10"

--------------------------
GCOV: OVERALL TEST SUMMARY
--------------------------
TESTED:  9
PASSED:  9
FAILED:  0
IGNORED: 0


---------------------------
GCOV: CODE COVERAGE SUMMARY
---------------------------
a.c Lines executed:100.00% of 27
a.c No branches
a.c Calls executed:100.00% of 3

b.c Lines executed:100.00% of 28
b.c No branches
b.c Calls executed:100.00% of 4

c.c Lines executed:100.00% of 28
c.c No branches
c.c Calls executed:100.00% of 4
```
</details>

<details>
  <summary>Run 2: Full console output</summary>

Notice how `Compiling a.c with coverage` appears **twice** in the logs and reported coverage for `a.c` is 33.33% at the end.

```
 VAGRANT|centos7|tester70  [2020-04-09][16:53:24] [/Volumes/.../git/ceedling_bug]
$ ceedling clobber gcov:all utils:gcov

Clobbering all generated files...
(For large projects, this task may take a long time to complete)



Test 'test_a.c'
---------------
Generating runner for test_a.c...
Compiling test_a_runner.c...
Compiling test_a.c...
Compiling unity.c...
Compiling a.c with coverage...
Compiling cmock.c...
Linking test_a.out...
Running test_a.out...


Test 'test_b.c'
---------------
Generating runner for test_b.c...
Compiling test_b_runner.c...
Compiling test_b.c...
Compiling b.c with coverage...
Linking test_b.out...
Running test_b.out...


Test 'test_c.c'
---------------
Generating runner for test_c.c...
Compiling test_c_runner.c...
Compiling test_c.c...
Compiling unity.c...
Compiling a.c with coverage...
Compiling c.c with coverage...
Compiling cmock.c...
Linking test_c.out...
Running test_c.out...
In your project.yml, define:

:gcov:
  :xml_report:

 to true or false to refine this feature.
For now, assumimg you do not want an xml report generated.
Creating a detailed html report of gcov results in build/artifacts/gcov/GcovCoverageResults.html...
Done.

-----------------
GCOV: TEST OUTPUT
-----------------
[test_a.c]
  - "a1() = 10"
  - "a2() = 10"
  - "a3() = 10"

[test_b.c]
  - "a1() = 10"
  - "b1() = 10"
  - "b2() = 10"
  - "b3() = 10"

[test_c.c]
  - "a1() = 10"
  - "b1() = 10"
  - "c1() = 10"
  - "c2() = 10"
  - "c3() = 10"

--------------------------
GCOV: OVERALL TEST SUMMARY
--------------------------
TESTED:  9
PASSED:  9
FAILED:  0
IGNORED: 0


---------------------------
GCOV: CODE COVERAGE SUMMARY
---------------------------
a.c Lines executed:33.33% of 27
a.c No branches
a.c Calls executed:33.33% of 3

b.c Lines executed:100.00% of 28
b.c No branches
b.c Calls executed:100.00% of 4

c.c Lines executed:100.00% of 28
c.c No branches
c.c Calls executed:100.00% of 4
```
</details>

I debugged the code and was able to trace what triggers recompilation down to this part of the code - https://github.com/ThrowTheSwitch/Ceedling/blob/master/lib/ceedling/task_invoker.rb#L96. In particular, inside `reset_rake_task_for_changed_defines`, `@project_config_manager.test_defines_changed` sometimes returns `TRUE` for source files that have been previously compiled which, if I understood correctly, triggers recompilation. Unfortunately I feel like I'm lacking better understanding of the big picture and how Ceedling decides which files and when to recompile to understand why that happens with `gcov:all` and not with `test:all`.

**Comment by CezaryGapinski** [comment]
Hi @IvanRibakov

Thank you for your detailed explanation and great example.

> I debugged the code and was able to trace what triggers recompilation down to this part of the code - https://github.com/ThrowTheSwitch/Ceedling/blob/master/lib/ceedling/task_invoker.rb#L96. In particular, inside `reset_rake_task_for_changed_defines`, `@project_config_manager.test_defines_changed` sometimes returns `TRUE` for source files that have been previously compiled which, if I understood correctly, triggers recompilation. 

Based on this I have found some issue when incorrect `TRUE` value can be returned from `diff_cached_defines` method. To correctly handle changed preprocessor defines I used `defines_dependency.yml` file with hash containing source files and current defines.

After compilation for `test_a` and `test_b` this file contains:
```yml
src/basic/a.c:
- TEST
src/basic/b.c:
- TEST
```

During `test_c` compilation the `defines_dependency.yml` do not yet contains new defines for `src/basic/c.c` required to compile`test_c`.

Current dependencies for the `test_c` compilation looks like this:
```yml
src/basic/a.c:
- TEST
src/basic/b.c:
- TEST
src/basic/c.c:
- TEST
```

but in the `defines_dependency.yml` we have only:
```yml
src/basic/a.c:
- TEST
src/basic/b.c:
- TEST
```

For this condition the `diff_cached_defines` return `TRUE` but should return `FALSE`. This method should compare intersection between current dependencies and already stored in the `defines_dependency.yml` to handle this type of situation correctly and do not recompile the `a.c` and `b.c` files. New source files shouldn't trigger recompilation the old ones when preprocessor defines do not change. I have made a [fix](https://github.com/CezaryGapinski/Ceedling/commit/a8e159f1bf62e1248d527e3257d0fd5ace2c9ec8) for this.

Can you apply this change for your example and confirm that helps to solve this issue?

**Comment by IvanRibakov** [comment]
Hi @CezaryGapinski ,

thanks a lot for looking into this so promptly! I can confirm that the fix you proposed solves the issue as demonstrated by the source code in the reproduction repo that I shared earlier.

As I applied the fix and ran it agains the real project that I'm working on, the issue seems to have persisted. Thanks to the fact that you broke down the problem really well in your previous comment I was able to compare the state of `defines_dependency.yml` file at various points throughout the ceedling run and realised that we've been using a custom `defines:` section for one of the source files:

```
:defines:
  :common: &common_defines
    - _GNU_SOURCE
    - TEST
    - SOURCE_PATH_SIZE=0
  :test: &test_defines
    - *common_defines
  :test_preprocess:
    - *common_defines
  :test_log: # defines specific for test_log.c
    - *test_defines
    - USE_REAL_LOGGER # needed to make sure that custom save-to-file logic is executed
```

Custom define for `test_log` test has been added a while ago and at that time we probably haven't been running ceedling with code coverage yet to notice the impact. 

Thankfully in our case this custom define is not a deal breaker and I think we can work around it in other ways, but it would be good to know if there is some way to run certain tests with custom defines and still get a complete code coverage picture.

**Comment by CezaryGapinski** [comment]
>Thankfully in our case this custom define is not a deal breaker and I think we can work around it in other ways, but it would be good to know if there is some way to run certain tests with custom defines and still get a complete code coverage picture.

I know that, but this is an another issue, similar to that one https://github.com/gcovr/gcovr/issues/338. Changed defines causes recompilation and this is desirable behaviour in Ceedling. But you have right, coverage for the same files with changed defines should be merged. This feature is already available in Gcovr thanks to `--add-tracefile` option, but this feature is not yet implemented in Ceedling. I have it on my TODO list :).

**Comment by IvanRibakov** [comment]
> I have it on my TODO list :).

Great, looking forward to testing it out!

Regarding your fix - can I assume it will be merged to Ceedling master soon?

**Comment by ojohnny** [comment]
The new release of Ceedling (0.30.0) has solved the problem for me. Should this be closed?

**Comment by mvandervoord** [comment]
:) It's fixed because of @CezaryGapinski 's fix above. Thanks everyone!
